### ISSUE TITLE
Automatic derivation with private implicits, "-Ywarn-unused" and Scala 2.12.4 produces erroneous warnings

### ISSUE THREAD

**Issue by migesok** [circe/circe#769]
Hi!
I discovered an issue after migration to Scala 2.12.4. It is possibly a bug in scalac but I'm not sure.
Here's minimal example:
From `build.sbt`:
```
scalaVersion := "2.12.4"
scalacOptions ++= Seq("-Ywarn-unused", "-Xfatal-warnings")
val circeVersion = "0.8.0"
libraryDependencies ++= Seq(
  "io.circe" %% "circe-core",
  "io.circe" %% "circe-generic",
  "io.circe" %% "circe-parser"
).map(_ % circeVersion)
```
`Test.scala`:
```
object Test extends App {
  import java.time.Instant
  import io.circe.generic.auto._
  import io.circe.syntax._
  
  private implicit val instantEncoder: io.circe.Encoder[Instant] =
    io.circe.Encoder.encodeLong.contramap(_.toEpochMilli)

  case class A(prop: Instant)

  println(A(Instant.now()).asJson)
}
```
Like this it fails to compile:
> private val instantEncoder in object Test is never used

If you comment out `instantEncoder` it fails also:
> could not find implicit value for parameter encoder: io.circe.Encoder[Test.A]

And it compiles fine with Scala 2.12.3.

**Comment by thatfulvioguy** [comment]
I've also ran into this issue.

This can be worked around in the meantime by annotating the offending declaration with [silent](https://github.com/ghik/silencer) or by change its visibility from `private` to `private[NameOfEnclosingType]`.

**Comment by javierarrieta** [comment]
We are also running into this issue, would love to help but don't know even where to start looking

**Comment by pdo400** [comment]
Any info on this or a link to somewhere where there is more discussion?

**Comment by pdo400** [comment]
Actually, just found these:
https://github.com/scala/bug/issues/10599
https://github.com/scala/bug/issues/10571

TL;DR:  Use the compiler flag `-Ywarn-macros:after`
